### PR TITLE
fix setMinZoom/setMaxZoom

### PR DESCRIPTION
**Pull request by cs09g**
fix setMinZoom/setMaxZoom method when view has resolutions property. 

#6746

- [x] This pull request addresses an issue that has been marked with the 'Pull request accepted' label & I have added the link to that issue.
- [x] It contains one or more small, incremental, logically separate commits, with no merge commits.
- [x] I have used clear commit messages.
- [x] Existing tests pass for me locally & I have added or updated tests for new or changed functionality.
- [x] The work herein is covered by a valid [Contributor License Agreement (CLA)](https://github.com/openlayers/cla).

### CODE REVIEW

File: src/ol/interaction/interaction.js [hunk 1]
```diff
@@ -180,6 +181,12 @@ ol.interaction.Interaction.zoom = function(view, resolution, opt_anchor, opt_dur
 ol.interaction.Interaction.zoomByDelta = function(view, delta, opt_anchor, opt_duration) {
   var currentResolution = view.getResolution();
   var resolution = view.constrainResolution(currentResolution, delta, 0);
+  var resolutions = view.getResolutions();
+
+  resolution = ol.math.clamp(
+      resolution,
```
Comment (cs09g) [inline src/ol/interaction/interaction.js]: it's very strict type check. should change `resolution || 0` for `resolution: undefined`?

Comment (ahocevar) [reply src/ol/interaction/interaction.js]: No, only do the clamp when `resolution !== undefined`.

File: src/ol/view.js [hunk 2]
```diff
@@ -815,7 +821,7 @@ ol.View.prototype.getZoomForResolution = function(resolution) {
   var max, zoomFactor;
   if (this.resolutions_) {
     var nearest = ol.array.linearFindNearest(this.resolutions_, resolution, 1);
-    offset += nearest;
+    offset = nearest;
```
Comment (ahocevar) [inline src/ol/view.js]: The `offset` variable is no longer used.

Comment (cs09g) [reply src/ol/view.js]: `offset` is used when it returns.

Comment (ahocevar) [reply src/ol/view.js]: Right :smile: - had missed that in the diff view.

**Comment by cs09g** [comment]
Thanks for the review!!

File: src/ol/interaction/interaction.js [hunk 3]
```diff
@@ -181,6 +182,14 @@ ol.interaction.Interaction.zoomByDelta = function(view, delta, opt_anchor, opt_d
   var currentResolution = view.getResolution();
   var resolution = view.constrainResolution(currentResolution, delta, 0);
 
+  if (resolution !== undefined) {
+    var resolutions = view.getResolutions();
+    resolution = ol.math.clamp(
+        resolution,
+        view.getMinResolution() || resolutions[resolutions.length - 1],
+        view.getMaxResolution() || resolutions[0]);
+  }
```
Comment (tschaub) [inline src/ol/interaction/interaction.js]: Previously, we would let people temporarily zoom outside of the min/max range.  This was intentional.  When using a trackpad for scrolling, for example, it feels nice to have the map temporarily zoom in/out beyond the constraints and then settle back to the constrained value.  This change breaks that behavior.

Comment (tschaub) [reply src/ol/interaction/interaction.js]: That might not be true...

Comment (tschaub) [reply src/ol/interaction/interaction.js]: Looks like this isn't the case.  Ignore the noise.

Comment (cs09g) [reply src/ol/interaction/interaction.js]: Thanks for the mention. I've got to check on mac device but without that condition, it doesn't work as expected for setMin/MaxZoom.

**Comment by waxenegger** [comment]
@cs09g and @tschaub 

I'd just like to point at one side-effect of this PR in the case when one wants to solely work with resolutions and hands no zoom bounds to the View. 

https://github.com/openlayers/openlayers/blob/v4.6.4/src/ol/view.js#L1113

In that case both, min/max resolution are forced to be the zoom max/min defined further up. What's more, if one happens to have lots of resolution levels the default value for min resolution will be clamped at index 28 of the resolutions (the default max for zoom levels)

**Comment by cs09g** [comment]
@waxenegger you're right about the second one. and when the zoom changed by `setZoom`, it can be moved to beyond min/max zoom. should be fixed.

**Pull request merged**